### ISSUE TITLE
Adjust for operation with Python 3.11 under Windows

### ISSUE THREAD

**Issue by peter88213** [vkbo/novelWriter#1284]
Actually, this is not a novelWriter bug. Anyway, I recommend to change the "shebang" from 
`#!/usr/bin/env python3`
to 
`#!/usr/bin/python3`
in order to avoid startup failures with the Python 3.11 launcher under Windows.

See: 
https://github.com/python/cpython/issues/100107

**Comment by vkbo** [comment]
Thanks for the heads up. This is really annoying since the recommended practice thus far has been the `env` method. I don't yet release novelWriter with 3.11 on Windows yet. Still on 3.10.

**Comment by vkbo** [comment]
I've been looking at this a bit. What is the use case here? Running novelWriter from source or from the installed version using the provided setup installer?

If it's the former, it requires changing the `novelWriter.py` file which is used for all platforms. In the latter case, it only involves modifying the generated `noverlWriter.pyw` file, which is simple enough to fix. However, in the latter case it is also designed to pick up the embedded pythonw.exe, not the system one.

**Comment by peter88213** [comment]
I think, this applies to running it from source, with a Python 3.11 installation from the python.org download page. 
I changed [my own installation script](https://github.com/peter88213/novelyst/blob/main/src/setup.pyw) to omit the shebang when building the startup script on Windows:
```
    #--- Create a start-up script.
    if platform.system() == 'Windows':
        shebang = ''
    else:
        shebang = '#!/usr/bin/python3\n'
    with open(f'{installDir}/{START_UP_SCRIPT}', 'w', encoding='utf-8') as f:
        f.write(f'{shebang}{START_UP_CODE}')
```
I do realize that this is not really necessary, because the new Python launcher works with `#!/usr/bin/python3`, but you never know what ideas the developers will come up with next when trying to emulate unix shell features under Windows.

**Comment by vkbo** [comment]
I'm not entirely sure what to do about this. I originally followed this recommendation when first moving beyond Linux for this project: https://stackoverflow.com/questions/6908143

Does it still launch with the old shebang if you call `python novelWriter.py`?

**Comment by peter88213** [comment]
`Does it still launch with the old shebang if you call python novelWriter.py?`

Yes, on the command line this seems to call directly the interpreter to which PATH leads. However, the point is to suppress the console window, so "pythonw" would be the better command. 
You can bypass the Python launcher in this way, but I have noticed that (for my *novelyst* program) starting by dragging and dropping a project file no longer works.

**Comment by vkbo** [comment]
Yes, I meant `pythonw`. Anyway, on the Windows installed version, I handle this by explicitly adding the pythonw.exe path bundled with novelWriter to the desktop and start menu icons. It passes novelWriter.pyw as a parameter instead.

https://github.com/vkbo/novelWriter/blob/66cc3777dc5072b5139013bfcd44e4edf5b83336/setup/win_setup_embed.iss#L48-L49

It was the only way I found to solve this without building an exe file of the project. I found that bundling everything in one large exe was both slow and sent half of the low end virus checkers panicking for no reason.

**Comment by peter88213** [comment]
On a side note, the *.pyw* extension seems to be meant as a discriminator for the Python launcher. If you call *pythonw.exe* directly, your script's extension may be *.py* as well. 

I have not yet dealt with package building, which means my distribution comes as a zipfile with Python scripts. Some of my setup scripts create Windows registry files for context menu entries or file name associations. Here the absolute path to the interpreter is required, so I get it from sys.argv[0] of the setup script, which therefore must also have the *.pyw* extension.

**Comment by vkbo** [comment]
The `.pyw` extension does no harm though and ensures that you can also launch it by double-click on the launch script, so it makes somewhat sense still.

Anyway, my packaging solution is to just add the embedded python downloadable from python.org in the build folder with the novelwriter source, and then run `pip` on the `requirements.txt` file targeting a `lib` folder at the same location. Then I inject that into position 0 in `sys.path` so that I have my own little isolated environment.

The resulting package is installed into a folder under Program Files or in user space, and runs as a normal uncompressed python application with the latest versions of Python 3.10 and dependencies. It is significantly snappier than any other solutions I've tried.

**Comment by peter88213** [comment]
That reminds me of the cordless tools you can buy either withBattery and charger, or "barebone". However, the full Python distribution is quite a chunk, isn't it? I can't offer this when providing a whole bunch of different Python tools. 
What about PyInstaller? As far as I understand it, it packs only the required modules. And I don't mean the "All-in-one-exe" solution you mentioned.

**Comment by vkbo** [comment]
I think I've tried them all, and they randomly set off the low end virus software because they tend to fingerprint the installers rather than the malware that uses those installers. It's the lazy, cheap approach.

The bulky dependency for novelWriter is PyQt5. I actually have a script that deletes two thirds of it though, as I don't use any of the bulky components like the web engine and various embedded bits. Neither do I use the QtQuick framework. The Windows installer is down to about 30 MB, which is manageable.

In comparison, the AppImage release file is about 90 MB, and the Debian release a little over 2 MB, which is what novelWriter really boils down to.

**Comment by peter88213** [comment]
So let's assume there are several useful applications, all programmed with Python and built on top of QT. Then the central Python installation would probably be the most economical solution, but also the one that places the most constraints on the applications. Presumably, this would be the choice of users who know their system reasonably well. In any case, we would then have the use case to which the "shebang problem" refers, to return to the topic.

**Comment by vkbo** [comment]
Indeed. For packages it isn't really an issue, as rewriting the files for the Windows packages is not a big deal and can be done in the `setup.py` script. The issue is still whether going back to the old style shebang is right for Linux and Mac, in terms of running from the source code.

**Comment by peter88213** [comment]
If you leave out the shebang on Windows, it doesn't matter what it looks like for the other operating systems.

**Comment by vkbo** [comment]
Was going through old bug tickets, and I'm wondering if this one is still relevant?

From next release, a novelWriter.exe will be distributed for Windows. It doesn't contain any app code, just a simple launcher that handles calling the correct python exe and startup script. So far it's worked on my VMs and a few users have tested it too.

**Comment by peter88213** [comment]
With Python version 3.12.8 I can start Python scripts that contain the `#!/usr/bin/env python3` shebang without problems. 
However, if it's not too much trouble, I would still recommend leaving out the “shebang” line in the Windows release. Does it still have any significance under Linux?

FYI:
If I'm not mistaken, the _Python Launcher_ included with the  Python 3.11 Windows installation tried to emulate `/usr/bin/env` under Windows, connecting with some Windows environment where the applications from the Microsoft Store are managed. This went wrong in case you got Python from another source, such as _python.org_ (which is, by the way, strongly recommended). It looks like they have fixed this with Python 3.12.

**Comment by vkbo** [comment]
> With Python version 3.12.8 I can start Python scripts that contain the `#!/usr/bin/env python3` shebang without problems. However, if it's not too much trouble, I would still recommend leaving out the “shebang” line in the Windows release.

Sure, the one in the Windows release is, as mentioned, auto-generated by the packaging script. With the new exe launcher, it has no use case any more as the user can just double-click the exe instead if trying to launch directly from the install folder.

> Does it still have any significance under Linux?

Yes, absolutely.

> FYI: If I'm not mistaken, the _Python Launcher_ included with the Python 3.11 Windows installation tried to emulate `/usr/bin/env` under Windows, connecting with some Windows environment where the applications from the Microsoft Store are managed. This went wrong in case you got Python from another source, such as _python.org_ (which is, by the way, strongly recommended). It looks like they have fixed this with Python 3.12.

novelWriter is distributed with Python embeddable from python.org on Windows, so that's a non-issue here. Currently shipping with 3.12.

I'll add this ticket to the 2.7 release then, so I can make the modifications mentioned here.

**Comment by peter88213** [comment]
As far as I'm concerned, you can close this issue. The “shebang” only affects people who run novelWriter under Windows via the sources and have Python 3.11 installed. 

> With the new exe launcher, it has no use case any more

Basically, the “shebang” should not have any meaning under Windows. It is only evaluated by the (optional) _Python Launcher_, which is intended to select the correct Python installation if there are several of them.
